app/services/fetch_service.rb

```diff
@@ -0,0 +1,451 @@
+    docs = docs.sort { |d1, d2| d1[:position].to_i <=> d2[:position].to_i }
+
+    docs.each_with_index do |doc, index|
+      doc[:explain] = doc[:explain].to_json if doc[:explain].is_a?(Hash)
+
+      doc_params = {
+        doc_id:     doc[:id],
+        explain:    doc[:explain],
+        position:   doc[:position] || (index + 1),
+        rated_only: doc[:rated_only] || false,
+        fields:     doc[:fields].blank? ? nil : doc[:fields].to_json,
+      }
+      results << query.snapshot_docs.build(doc_params)
+    end
+
+    results
+  end
+  # rubocop:enable Metrics/MethodLength
+  # rubocop:enable Metrics/CyclomaticComplexity
+  # rubocop:enable Metrics/PerceivedComplexity
+
+  # This maybe should be split out into a snapshot_query and a snapshot_docs?
+  def store_query_results query, docs, response_status, response_body
+    snapshot_query = @snapshot.snapshot_queries.create(
+      query:             query,
+      number_of_results: docs.count,
+      response_status:   response_status
+    )
+    snapshot_query.create_web_request(
+      response_status: response_status,
+      response:        response_body
+    )
+    snapshot_manager = SnapshotManager.new(@snapshot)
+    query_docs = snapshot_manager.setup_docs_for_query(snapshot_query, docs)
+    SnapshotDoc.import query_docs
+
+    # if we don't need the snapshot_docs data, maybe we don't need the reload??
+    # snapshot_query.reload
+    snapshot_query
+  end
+
+  def score_run
+    @snapshot.name = 'Fetch [PREPARE_SCORE]'
+    @snapshot.save!
+    score_data = score_snapshot @snapshot, @try
+
+    case_score_manager = CaseScoreManager.new @case
+    case_score_manager.update score_data
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  # rubocop:disable Metrics/BlockLength
+  # rubocop:disable Layout/LineLength
+  def score_snapshot snapshot, try
+    queries_detail = {}
+
+    snapshot.snapshot_queries.each do |snapshot_query|
+      javascript_scorer = JavascriptScorer.new(Rails.root.join('db/scorers/scoring_logic.js'))
+
+      # Prepare some items to score
+      doc_ratings = {}
+      snapshot_query.query.ratings.each do |rating|
+        doc_ratings[rating.doc_id] = rating.rating
+      end
+
+      docs = snapshot_query.snapshot_docs.map do |snapshot_doc|
+        { id: snapshot_doc.doc_id, rating: doc_ratings[snapshot_doc.doc_id] }
+      end
+
+      best_docs = snapshot_query.query.ratings.map do |rating|
+        { id: rating.doc_id, rating: rating.rating }
+      end
+
+      best_docs.sort_by! { |doc| doc[:rating] }.reverse
+
+      # docs = [
+      #  { id: 1, value: 10, rating: 3 },
+      #  { id: 2, value: 20, rating: 0 }
+      # ]
+
+      # Calculate score with options
+      begin
+        scorer = snapshot.scorer
+        code = scorer.code
+
+        score = javascript_scorer.score(docs, best_docs, code)
```

Comment:
@david-fisher we sometimes get NaN back from a scorer.  Is this okay?   Right now the logic is to force that to a zero.  That is also how the Angular app works too...   Or, do we fix it now?